File: iosApp/iosApp/Sync/SyncViewModel.swift
//
//  SyncViewModel.swift
//  iosApp
//
//  Created by Kwabena Berko on 25/11/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import shared
import KMPNativeCoroutinesAsync

final class SyncViewModel: BaseViewModel<SyncViewModel.State> {
    private var sync: Sync
    
    init(sync: Sync){
        self.sync = sync
        super.init(initialState: .idle)
        
        startSync()
    }
    
    func startSync() {
        Task {
            setState(newState: .syncing)
            
            do {
                let isSuccess = try await asyncFunction(for: sync.invokeNative()).boolValue
                if(!isSuccess){
                    setState(newState: .syncError)
                } else {
                    setState(newState: .syncSuccess)
                }
            } catch {
                print("Failed with error: \(error)")
            }
        }
    }
    
    enum State {
        case idle
        case syncing
        case syncError
        case syncSuccess
    }
}

File: shared/src/commonMain/kotlin/com/kwabenaberko/converter/presentation/viewmodel/SyncViewModel.kt
package com.kwabenaberko.converter.presentation.viewmodel

import com.kwabenaberko.converter.domain.usecase.Sync
import kotlinx.coroutines.launch

class SyncViewModel(
    private val sync: Sync
) : BaseViewModel<SyncViewModel.State>(State.Idle) {

    init {
        startSync()
    }

    fun startSync() {
        scope.launch {

            setState(State.Syncing)

            when (sync()) {
                true -> setState(State.SyncSuccess)
                false -> setState(State.SyncError)
            }
        }
    }

    sealed class State {
        object Idle : State()
        object Syncing : State()
        object SyncError : State()
        object SyncSuccess : State()
    }
}
